From patchwork Mon Oct  8 18:06:20 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Wenwen Wang <wang6495@umn.edu>
X-Patchwork-Id: 10631149
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 41E9114BD
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  8 Oct 2018 18:06:41 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 3782429980
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  8 Oct 2018 18:06:41 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 2BB85299A7; Mon,  8 Oct 2018 18:06:41 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI autolearn=ham
	version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id C774529980
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  8 Oct 2018 18:06:40 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726731AbeJIBTb (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 8 Oct 2018 21:19:31 -0400
Received: from mta-p7.oit.umn.edu ([134.84.196.207]:47266 "EHLO
        mta-p7.oit.umn.edu" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726712AbeJIBTa (ORCPT <rfc822;kvm@vger.kernel.org>);
        Mon, 8 Oct 2018 21:19:30 -0400
Received: from localhost (unknown [127.0.0.1])
        by mta-p7.oit.umn.edu (Postfix) with ESMTP id BFC1E1049
        for <kvm@vger.kernel.org>; Mon,  8 Oct 2018 18:06:34 +0000 (UTC)
X-Virus-Scanned: amavisd-new at umn.edu
Received: from mta-p7.oit.umn.edu ([127.0.0.1])
        by localhost (mta-p7.oit.umn.edu [127.0.0.1]) (amavisd-new,
 port 10024)
        with ESMTP id oC2a1IoaMY6P for <kvm@vger.kernel.org>;
        Mon,  8 Oct 2018 13:06:34 -0500 (CDT)
Received: from mail-it1-f197.google.com (mail-it1-f197.google.com
 [209.85.166.197])
        (using TLSv1.2 with cipher AES128-GCM-SHA256 (128/128 bits))
        (No client certificate requested)
        by mta-p7.oit.umn.edu (Postfix) with ESMTPS id 846C91014
        for <kvm@vger.kernel.org>; Mon,  8 Oct 2018 13:06:34 -0500 (CDT)
Received: by mail-it1-f197.google.com with SMTP id y142-v6so12058929itb.7
        for <kvm@vger.kernel.org>; Mon, 08 Oct 2018 11:06:34 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=umn.edu; s=google;
        h=from:to:cc:subject:date:message-id;
        bh=NW2oLiXs1xkho6nwIyy1q2qHL+VNYiwokideRStX/Rs=;
        b=PJMh5AHZByqhxGK0o6MyMfQktOijHi9jel47BnnVCpDfSKVwpVVYt6EJwyaIYfVQta
         fwAVuaRCPh54jhWSPLjvBOYMqKlkjQu5XSSRtsJIwgPP+1YPeq/gpas892Bc/lNXVPmS
         guIDMGUj4nxeuloQu3VDFI7/DpYLint5ftlglf1iVd5C9r7JuKPBOa9QYro+8K6hz/F5
         irnyaVKrOXWiCWKQqYszR2C/GKY8Gb4PQcmHEhIHlbZ7oZrtCJhTejvNlJs2GUYoq/wz
         Wig+90WFN/d3ZGiNS8lYO3/DwuZTEgqdEqU8s8e9Ko82sMajdSirhaEPHcwTABsYdFES
         woxQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id;
        bh=NW2oLiXs1xkho6nwIyy1q2qHL+VNYiwokideRStX/Rs=;
        b=QgxSYhvGASNQWKu0zAL0usajuCEMutJMKjFPvuvgwcKXLVgciQh16NMjvg6RL4QVOk
         vnEBDmY5ziYFIJTsy1ktoibEro8Vkgv9ceznspb7dDrc+1VzLWPALm02/DBDEi4j+/wP
         nJy9wd5HZ8KEk/2JjNsLl2sMTgAgYhfRpD5TwvnszrScc6nyUK6Fa6uQO4oEFDqmxmCp
         jYagIxfdiH0H5b7k+5/pbsC4KUssIf/2rhRCxrL51Li7jofruW+FkriZp5TiTSQvGBv4
         RPMEFWlTjXjeopa/vE2iGp+BPtTXs/omNb8Q8z5QMkPYn+7ze2F3a7VYRw0QAOXoxCHb
         ATcw==
X-Gm-Message-State: ABuFfojO40FWo8t5CQdRytk/j7I6/98QsQ9+arr8J6/o5RQSfx2x9KN8
        yNxoom8jZ7pvFoFOjmdycWMJpiNmVs4scfJyilQEuc66aB/UcRDrXGcetLpH353v7KIUhbPAtze
        7Vb2+MTjkKR7w6SsJ
X-Received: by 2002:a02:91c9:: with SMTP id
 s9-v6mr20561383jag.104.1539021994187;
        Mon, 08 Oct 2018 11:06:34 -0700 (PDT)
X-Google-Smtp-Source: 
 ACcGV63HrYkFMF6Pg9CRs+2l6z+F5YNMDuJ4p+6u/AYeM9F0+8mx3ThYHkzMjAnatFl/780yF+VUWg==
X-Received: by 2002:a02:91c9:: with SMTP id
 s9-v6mr20561373jag.104.1539021993995;
        Mon, 08 Oct 2018 11:06:33 -0700 (PDT)
Received: from cs-u-cslp16.cs.umn.edu (cs-u-cslp16.cs.umn.edu.
 [134.84.121.95])
        by smtp.gmail.com with ESMTPSA id
 p185-v6sm450411itg.34.2018.10.08.11.06.32
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-SHA bits=128/128);
        Mon, 08 Oct 2018 11:06:33 -0700 (PDT)
From: Wenwen Wang <wang6495@umn.edu>
To: Wenwen Wang <wang6495@umn.edu>
Cc: Kangjie Lu <kjlu@umn.edu>,
        Alex Williamson <alex.williamson@redhat.com>,
        kvm@vger.kernel.org (open list:VFIO DRIVER),
        linux-kernel@vger.kernel.org (open list)
Subject: [PATCH v2] drivers/vfio: Fix a redundant copy bug
Date: Mon,  8 Oct 2018 13:06:20 -0500
Message-Id: <1539021980-2412-1-git-send-email-wang6495@umn.edu>
X-Mailer: git-send-email 2.7.4
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

In vfio_spapr_iommu_eeh_ioctl(), if the ioctl command is VFIO_EEH_PE_OP,
the user-space buffer 'arg' is copied to the kernel object 'op' and the
'argsz' and 'flags' fields of 'op' are checked. If the check fails, an
error code EINVAL is returned. Otherwise, 'op.op' is further checked
through a switch statement to invoke related handlers. If 'op.op' is
VFIO_EEH_PE_INJECT_ERR, the whole user-space buffer 'arg' is copied again
to 'op' to obtain the err information. However, in the following execution
of this case, the fields of 'op', except the field 'err', are actually not
used. That is, the second copy has a redundant part. Therefore, for both
performance consideration, the redundant part of the second copy should be
removed.

This patch removes such a part in the second copy. It only copies from
'err.type' to 'err.mask', which is exactly required by the
VFIO_EEH_PE_INJECT_ERR op.

Signed-off-by: Wenwen Wang <wang6495@umn.edu>
---
 drivers/vfio/vfio_spapr_eeh.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/drivers/vfio/vfio_spapr_eeh.c b/drivers/vfio/vfio_spapr_eeh.c
index 38edeb4..66634c6 100644
--- a/drivers/vfio/vfio_spapr_eeh.c
+++ b/drivers/vfio/vfio_spapr_eeh.c
@@ -37,6 +37,7 @@ long vfio_spapr_iommu_eeh_ioctl(struct iommu_group *group,
 	struct eeh_pe *pe;
 	struct vfio_eeh_pe_op op;
 	unsigned long minsz;
+	unsigned long start, end;
 	long ret = -EINVAL;
 
 	switch (cmd) {
@@ -86,10 +87,12 @@ long vfio_spapr_iommu_eeh_ioctl(struct iommu_group *group,
 			ret = eeh_pe_configure(pe);
 			break;
 		case VFIO_EEH_PE_INJECT_ERR:
-			minsz = offsetofend(struct vfio_eeh_pe_op, err.mask);
-			if (op.argsz < minsz)
+			start = offsetof(struct vfio_eeh_pe_op, err.type);
+			end = offsetofend(struct vfio_eeh_pe_op, err.mask);
+			if (op.argsz < end)
 				return -EINVAL;
-			if (copy_from_user(&op, (void __user *)arg, minsz))
+			if (copy_from_user(&op.err, (char __user *)arg +
+						start, end - start))
 				return -EFAULT;
 
 			ret = eeh_pe_inject_err(pe, op.err.type, op.err.func,
